File: src/MyPrayerJournal/Entities.fs
namespace MyPrayerJournal

open Newtonsoft.Json

/// A user
type User = {
  /// The Id of the user
  [<JsonProperty("id")>]
  Id : string
  /// The user's e-mail address
  Email : string
  /// A hash of the user's password
  PasswordHash : string
  /// The user's name
  Name : string
  /// The last time the user logged on
  LastSeenOn : int64
}

/// Request history entry
type History = {
  /// The instant at which the update was made
  AsOf : int64
  /// The action that was taken on the request
  Action : string list
  /// The status of the request (filled if it changed)
  Status : string option
  /// The text of the request (filled if it changed)
  Text : string option
}

/// A prayer request
type Request = {
  /// The Id of the request
  [<JsonProperty("id")>]
  Id : string
  /// The Id of the user to whom this request belongs
  UserId : string
  /// The instant this request was entered
  EnteredOn : int64
  /// The history for this request
  History : History list
}
  with
    /// The current status of the prayer request
    member this.Status =
      this.History
      |> List.sortBy (fun item -> -item.AsOf)
      |> List.map    (fun item -> item.Status)
      |> List.filter Option.isSome
      |> List.map    Option.get
      |> List.head
    /// The current text of the prayer request
    member this.Text =
      this.History
      |> List.sortBy (fun item -> -item.AsOf)
      |> List.map    (fun item -> item.Text)
      |> List.filter Option.isSome
      |> List.map Option.get
      |> List.head
    member this.LastActionOn =
      this.History
      |> List.sortBy (fun item -> -item.AsOf)
      |> List.map    (fun item -> item.AsOf)
      |> List.head
      

File: src/api/data/entities.go
package data

// History is a record of action taken on a prayer request, including updates to its text.
type History struct {
	RequestID string `json:"requestId"`
	AsOf      int64  `json:"asOf"`
	Status    string `json:"status"`
	Text      string `json:"text"`
}

// Note is a note regarding a prayer request that does not result in an update to its text.
type Note struct {
	RequestID string `json:"requestId"`
	AsOf      int64  `json:"asOf"`
	Notes     string `json:"notes"`
}

// Request is the identifying record for a prayer request.
type Request struct {
	ID        string `json:"requestId"`
	EnteredOn int64  `json:"enteredOn"`
	UserID    string `json:"userId"`
}

// JournalRequest is the form of a prayer request returned for the request journal display. It also contains
// properties that may be filled for history and notes.
type JournalRequest struct {
	RequestID  string    `json:"requestId"`
	UserID     string    `json:"userId"`
	Text       string    `json:"text"`
	AsOf       int64     `json:"asOf"`
	LastStatus string    `json:"lastStatus"`
	History    []History `json:"history"`
	Notes      []Note    `json:"notes"`
}
